File: Projects/po_projekt2/organisms/Animal.java
package po_projekt2.organisms;

import java.awt.Color;
import po_projekt2.Organism;
import po_projekt2.World;
import po_projekt2.organisms.animals.Human;
import po_projekt2.worldElements.Coordinates;

/**
 *
 * @author Bogna
 */
public abstract class Animal extends Organism{
    protected int moveRange;
    
    public Animal(int strenght, int initiative, int  whenWasBorn, int moveRange, Color color, String name, Coordinates coords, World world, typeOfOrganism type, boolean hasSpecialColision){
        super(strenght, initiative, whenWasBorn, color, name, coords, hasSpecialColision, true, world, type);
        this.moveRange = moveRange;
    }
    
    @Override
    public void action(){
	int x = this.getCoords().getX();
	int y = this.getCoords().getY();
	this.setDirections(moveRange);
	direction d = this.randDirection();
	if (d == direction.UP) {
            y -= moveRange;
	}else if (d == direction.DOWN) {
            y += moveRange;
	}else if(d == direction.LEFT){
            x += moveRange;
	}else if (d == direction.RIGHT) {
            x -= moveRange;
	}else {
            return;
	}
	Organism tmp = world.getBoard().getOrganism(x, y);
	if (tmp == null) {
            Coordinates c1 = this.getCoords();
            this.setCoords(x, y);
            world.getBoard().moveOrganism(this, c1);
	}else {
            this.colision(this, tmp, x, y);
	}
    }
    
    @Override
    public void colision(Organism attacking, Organism other, int x, int y){
        if(other.getType() == attacking.getType()){
            this.reproduction((Animal)attacking, (Animal)other);
        }else if(other.getHasSpecialColision()){
            other.specialColision(attacking, x, y, false);
        }else if(attacking.getHasSpecialColision()){
            this.specialColision(other, x, y, true);
        }else if(attacking.getStrenght() >= other.getStrenght()){
            if(other instanceof Human){
                attacking.getWorld().setIsHumanAlive(false);
            }
            attacking.getWorld().deleteOrganism(other);
            Coordinates c1 = this.getCoords();
	    attacking.setCoords(x, y);
            attacking.getWorld().getBoard().moveOrganism(attacking, c1);
            world.addComment(attacking.getOrganismName() + "[x = " + attacking.getCoords().getX() + "; y = " + attacking.getCoords().getY() + "] kills " + other.getOrganismName() + "[x = " + other.getCoords().getX() + "; y = " + other.getCoords().getY() + ".\n");
        }else{
            if (attacking instanceof Human) {
                    other.getWorld().setIsHumanAlive(false);
            }
            other.getWorld().getBoard().removeOrganism(attacking.getCoords());
            other.getWorld().deleteOrganism(attacking);
            world.addComment(other.getOrganismName() + "[x = " + other.getCoords().getX() + "; y = " + other.getCoords().getY() + "] kills " + attacking.getOrganismName() + "[x = " + attacking.getCoords().getX() + "; y = " + attacking.getCoords().getY() + ".\n");
        }
    }
    
    @Override
    public void specialColision(Organism other, int x, int y, boolean isAttacking){}
    
    public int getMoveRange(){
        return this.moveRange;
    }
    
    protected void reproduction(Animal moving, Animal other){
        if (moving.reproduced || other.reproduced) {
            return;
	}else{
            Coordinates newCoords = new Coordinates();
            setFreeDirections(1);
            direction d = randDirection();
            int x = getCoords().getX(), y = getCoords().getY();
            if (d == direction.UP) {
                y -= 1;
            }else if (d == direction.DOWN) {
                y += 1;
            }else if (d == direction.LEFT) {
                x += 1;
            }else if (d == direction.RIGHT) {
                x -= 1;
            }else {
                return;
            }
            newCoords.setX(x);
            newCoords.setY(y);
            Organism newOrganism = world.createNewOrganism(moving.getType(), newCoords);
            moving.getWorld().addOrganism(newOrganism);
            moving.getWorld().getBoard().setOrganism(newOrganism);
            moving.setReproduced(true);
            other.setReproduced(true);
            world.addComment("New organism was born: " + newOrganism.getOrganismName() + "[ x = " + newOrganism.getCoords().getX() + "; y = " + newOrganism.getCoords().getY() + "].\n");
	}
    }
    
}

File: Project_1-c++/Animal.cpp
#include "Animal.h"
#include "Human.h"

Animal::Animal(int strenght, int initiative, int  whenWasBorn, int moveRange, char symbol, Coordinates coords, World* world, TypeOfOrganism type, bool hasSpecialColision) : Organism(strenght, initiative, whenWasBorn, symbol, coords, world, type, hasSpecialColision, true){
	this->moveRange = moveRange;
}

void Animal::reproduction(Animal* moving, Animal* other) {
	if (moving->reproduced || other->reproduced) {
		return;
	}else{
		Coordinates newCoords;
		SetFreeDirections(1);
		Direction d = randDirection();
		int x = GetCoordinates().GetX(), y = GetCoordinates().GetY();
		if (d == Direction::UP) {
			y -= 1;
		}else if (d == Direction::DOWN) {
			y += 1;
		}else if (d == Direction::LEFT) {
			x += 1;
		}else if (d == Direction::RIGHT) {
			x -= 1;
		}else {
			return;
		}
		newCoords.SetX(x);
		newCoords.SetY(y);
		Organism* newOrganism = world->createNewOrganism((int)moving->type, newCoords);
		moving->GetWorld()->addOrganism(newOrganism);
		moving->GetWorld()->GetBoard()->SetOrganism(newOrganism);
		moving->SetReproduced(true);
		other->SetReproduced(true);
		world->printMessage(newOrganism, other, false, false);
	}
}

void Animal::colision(Organism* attacking, Organism* other, int x, int y) {
	if (other->GetType() == type) {
		this->reproduction(this, dynamic_cast<Animal*>(other));
	}else if (other->GetHasSpecialColision()) {
		other->specialColision(attacking, x, y, false);
	}else if (this->GetHasSpecialColision()) {
		attacking->specialColision(other, x, y, true);
	}else if(attacking->GetStrenght() >= other->GetStrenght()) {
		world->printMessage(attacking, other, true, false);
		if (dynamic_cast<Human*>(other)) {
			world->SetIsHumanAlive(false);
		}
		attacking->GetWorld()->deleteOrganism(other);
		Coordinates c1 = this->GetCoordinates();
		attacking->SetCoorinates(x, y);
		world->GetBoard()->moveOrganism(attacking, c1);
		delete other;
	}else{
		world->printMessage(other, attacking, true, false);
		if (dynamic_cast<Human*>(attacking)) {
			world->SetIsHumanAlive(false);
		}
		world->GetBoard()->removeOrganism(attacking);
		other->GetWorld()->deleteOrganism(attacking);
		delete attacking;
	}
}

void Animal::specialColision(Organism* other, int x, int y, bool isAttacking) {}

void Animal::action() {
	Coordinates c;
	int x = this->GetCoordinates().GetX();
	int y = this->GetCoordinates().GetY();
	this->SetDirections(moveRange);
	Direction d = this->randDirection();
	if (d == Direction::UP) {
		y -= moveRange;
	}else if (d == Direction::DOWN) {
		y += moveRange;
	}else if(d == Direction::LEFT){
		x += moveRange;
	}else if (d == Direction::RIGHT) {
		x -= moveRange;
	}else {
		return;
	}
	c.SetX(x);
	c.SetY(y);
	Organism* tmp = world->GetBoard()->GetOrganism(c);
	if (tmp == nullptr) {
		Coordinates c1 = this->GetCoordinates();
		this->SetCoorinates(x, y);
		world->GetBoard()->moveOrganism(this, c1);
	}else {
		this->colision(this, tmp, x, y);
	}
};

int Animal::GetMoveRange() {
	return moveRange;
}

Animal::~Animal() {}

